File: 445-Add Two Numbers II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        string s1, s2;
        
        while (l1){
            s1 += to_string(l1->val);
            l1 = l1->next; 
        }
        while (l2){
            s2 += to_string(l2->val);
            l2 = l2->next;
        }
        int x = stoi(s1) + stoi(s2);
        string s3 = to_string(x);
        ListNode *dummy = new ListNode(-1);
        ListNode *cur = dummy;
        
        for(int i = 0; i < s3.size(); i++){
            int val = int(s3[i] - '0');
            cur->next = new ListNode(val);
            //cout<<cur->val<<endl;; 
            cur = cur->next;
        }
        
        return dummy->next;
    }
};

File: 445-Add Two Numbers II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        s1 = []
        while l1:
            s1.append(str(l1.val))
            l1 = l1.next
        
        s2 = []
        while l2:
            s2.append(str(l2.val))
            l2 = l2.next
        
        x = int(''.join(s1)) + int(''.join(s2))
        
        cur = dummy = ListNode(-1)
        for i in str(x):
            cur.next = ListNode(i)
            cur = cur.next
        
        return dummy.next
            
        
        
    def addTwoNumbers(self, l1, l2):
        def reverse(node):
            prev = None
            x = node
            while x:
                y = x.next
                x.next = prev
                prev = x
                x = y
            
            return prev
        l1 = reverse(l1)
        l2 = reverse(l2)
        head = None
        carry = digit = 0
        while l1 or l2:
            v1 = l1.val if l1 else 0
            v2 = l2.val if l2 else 0
            digit = (carry + v1 + v2) % 10
            carry = (carry + v1 + v2) // 10
            
            cur = ListNode(digit)
            cur.next = head
            head = cur
            #print(head)
            
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else 0
        
        if carry:
            cur = ListNode(carry)
            cur.next = head
            head = cur
        
        return head
            
            